### ISSUE TITLE
502 error when SESSION_ENGINE is signed_cookies

### ISSUE THREAD

**Issue by amir-qayyum-khan** [wagtail/wagtail#4521]
### Issue Summary

We are trying to preview a page in cms. Preview are store in session unlike drafts. Because description text is huge we are getting 502 error.

```Cookie length should be less than or equal to 4096 characters.```

### Steps to Reproduce

1. Set `SESSION_ENGINE = 'django.contrib.sessions.backends.signed_cookies'` in settings.py
2. login to cms
3. go tp pages
4. create a new child page or edit existing.
5. Add big text in description i.e [sample text](https://github.com/mitodl/micromasters/issues/3453#issuecomment-320938670)
6. click preview

@pdpinch
#### Screens:
<img width="1057" alt="screen shot 2018-05-08 at 12 02 07 am" src="https://user-images.githubusercontent.com/10431250/39719587-18c6f5b4-5253-11e8-9599-630a99ad058b.png">

<img width="1209" alt="screen shot 2018-05-08 at 12 01 52 am" src="https://user-images.githubusercontent.com/10431250/39719585-185edbb4-5253-11e8-85d0-2d4516f6738d.png">

<img width="1171" alt="screen shot 2018-05-08 at 12 02 01 am" src="https://user-images.githubusercontent.com/10431250/39719586-18944c2c-5253-11e8-98a0-bcaf2475386d.png">

### Technical details

* Python version: 3.6
* Django version: 2.0.2
* SESSION_ENGINE = ['django.contrib.sessions.backends.cached_db'
](https://docs.djangoproject.com/en/2.0/topics/http/sessions/#using-cached-sessions)
* Wagtail version: 2.0
* Browser version: chrome Version 66.0.3359.139 (Official Build) (64-bit) osx

**Comment by loicteixeira** [comment]
Thanks for the report @amir-qayyum-khan. 

For reference, preview data is stored in the session since #3383. While #3583 reduces preview data size, the issue will essentially still exists.

The fact that it might exceed the cookies max-size was raised at the time but it was agreed to add a warning to the documentation (although it seems to have never happened) given that not using the session probably meant creating an API endpoint which would have been a lot more work. Maybe it will be time to revisit this.

In the meantime, you should be able to configure uwsgi to accept bigger cookies (IRRC, nginx has some sort of auto-scaling but uwsgi doesn't).

**Comment by amir-qayyum-khan** [comment]
thanks @loicteixeira for you feedback.

We already have uwsgi config https://github.com/mitodl/micromasters/blob/master/uwsgi.ini#L47. But still we are getting preview error.

**Comment by blarghmatey** [comment]
@loicteixeira is there any particular reason not to treat previews as an as-yet unpublished revision? it seems as though that would be the most robust option for managing data that you would like to preview and eventually publish, though I'm happy to be proven wrong if there is something that I'm missing.

**Comment by gasman** [comment]
@blarghmatey By definition, I think a preview has to show the outcome of a proposed change without committing that change to the database (even just as a draft revision). If an editor hits preview, and then changes their mind about their edit, and navigates away without saving, the system should behave as if the edit was never made at all (and thus revisiting the edit view should take you back to the page's original state).

**Comment by loicteixeira** [comment]
Agreed that at the moment, it doesn't seem right to save a revision for a preview.

Once revision data is decoupled from page data with [RFC29](https://github.com/wagtail/rfcs/blob/master/text/029-immutable-revision-events.md), I imagine it would be possible to save a revision for the sole purpose of previewing as it would be possible to "tag" a revision as preview and exclude it from the list of revisions the user can look at (to reduce noise) and/or make sure a "preview draft" always deletes the previous one (so we don't overload the DB). I'm not sure that what we would like to do though, but it would certainly be less of a problem than with the current implementation of revisions.

In the meantime, if setting up uswgi and nginx doesn't help, the [session engine](https://docs.djangoproject.com/en/2.0/ref/settings/#std:setting-SESSION_ENGINE) could be set to anything but cookies (although it might not be possible for all projects).

**Comment by lb-** [comment]
RFC 29 was not adopted. 

https://github.com/wagtail/rfcs/blob/master/text/029-immutable-revision-events.md

Updated title to include Preview and adding labels.

**Comment by mgax** [comment]
As a work-around, it is possible to configure the preview mechanism to expire old data faster than the default 24 hours, by adding this to `wagtail_hooks.py`:

```python
from wagtail.admin.views.generic.preview import PreviewOnEdit

# Override expiration of preview data from 24 hours to 10 seconds
PreviewOnEdit.preview_expiration_timeout = 10
```
